### ISSUE TITLE
Doesn't install on Ruby 2.5

### ISSUE THREAD

**Issue by pboling** [janlelis/ruby_engine#1]
This *only* happens on Ruby 2.5.  It installs fine on all other versions.
```
ArgumentError: wrong number of arguments (given 4, expected 1)
An error occurred while installing ruby_engine (2.0.1), and Bundler cannot
continue.
Make sure that `gem install ruby_engine -v '2.0.1' --source
'https://rubygems.org/'` succeeds before bundling.
```

Ref: https://github.com/pboling/sanitize_email/actions/runs/8741467464/job/23987701059

**Comment by janlelis** [comment]
Hi Peter, that's a strange error indeed... The only code that has changed is this [workaround in the Gemfile](https://github.com/janlelis/ruby_engine/blob/d6d8ef846ee0cdeb2badef76c8c9af0e1b378b98/Gemfile#L3-L4) and maybe some clean-up in the Rakefile.

Will it help if I yank v2.0.1 and release v2.0.2 with a Ruby requirement of `>= 2.6`?

**Comment by pboling** [comment]
No, I don't think it is worth yanking, and I am using this in a gem that is supporting back to Ruby 2.3, so it wouldn't help my situation at all.  I actually just took Ruby 2.5 out of the CI matrix, as it *is working* on Ruby 2.3 and 2.4, so no need to rush a fix on my behalf.

**Comment by pboling** [comment]
Separately, why are you pegged to RSpec 2.99?  Need help with the syntax transition?  I could submit a PR.  I have a lot of experience with that transition, and then you could trash that hack.

**Comment by pboling** [comment]
Also, this is the strangest line of Ruby code I've ever seen, I think.

```
def File.exists?(f)exist?(f)end unless defined? File.exists?
```

I'm shocked that it works.

**Comment by pboling** [comment]
I'd guess that separating to multiline, and using parens with `defined?(File.exists?)` would fix the issue.

**Comment by pboling** [comment]
I'm guessing here, as I can't run 2.5 locally on the current machine, but I think what is happening in this error is `def` is receiving all four tokens as individual arguments to `def`, where it normally just expects one: `def method_name`.
The arguments it is receiving are:

1. `File.exists?(f)exist?(f)end`
2. `unless`
3. `defined?`
4. `File.exists?`

Either that, or `File.exists?(f)` is receiving the following tokens as additional arguments?  In any case, 2.5 doesn't handle that single line leet code at all.

**Comment by janlelis** [comment]
Hey! Cannot test it either, but I went for the "proper" fix and unpegged RSpec. Didn't need to change anything, but I could remove the leet code 😅

Could you give a feedback if the gem is now installable?

**Comment by pboling** [comment]
Interesting.  It still fails.  I've just noticed that it is actually failing on both `ruby_engine` and `ruby_version`, and that they are both failing with the same error, and only on Ruby 2.5.

I'm at a loss.  I'll look into this a bit more soon, after I finish up some other stuff.

```
ArgumentError: wrong number of arguments (given 4, expected 1)
An error occurred while installing ruby_engine (2.0.2), and Bundler cannot
continue.
Make sure that `gem install ruby_engine -v '2.0.2' --source
'https://rubygems.org/'` succeeds before bundling.

In vanilla.gemfile:
  rspec-pending_for was resolved to 0.1.16, which depends on
    ruby_engine


ArgumentError: wrong number of arguments (given 4, expected 1)
An error occurred while installing ruby_version (1.0.3), and Bundler cannot
continue.
Make sure that `gem install ruby_version -v '1.0.3' --source
'https://rubygems.org/'` succeeds before bundling.

In vanilla.gemfile:
  rspec-pending_for was resolved to 0.1.16, which depends on
    ruby_version
Error: The process '/opt/hostedtoolcache/Ruby/2.5.9/x64/bin/bundle' failed with exit code 5
```

ref: https://github.com/pboling/sanitize_email/actions/runs/8783962216/job/24101214809

**Comment by pboling** [comment]
I noticed you mentioned cleanup in the Rakefile, and I'm now thinking that's where the issue is.

All of this is wrong:
```
require 'rubygems'

begin
  require 'bundler'
rescue LoadError => e
  warn e.message
  warn "Run `gem install bundler` to install Bundler."
  exit -1
end

begin
  Bundler.setup(:development)
rescue Bundler::BundlerError => e
  warn e.message
  warn "Run `bundle install` to install missing gems."
  exit e.status_code
end

require 'rake'

require 'rubygems/tasks'
Gem::Tasks.new
```

Rakefiles can only be run by `rake` and should never load `rake`, as `rake` is already loaded and running when the file is read.  What is is loaded by?  rubygems!  So that shouldn't be required either.  Additionally, it makes no sense to care if bunder is present or not.  If doing `Bundler.setup` is to ensure certain gems are loaded when running specs, all those gems should be instead explicitly required at the top of the `spec_helper.rb`.

I have a similarly old gem, which I've recently modernized, the one whose build is failing due to this issue, and here is the Rakefile, if it helps:

```
# frozen_string_literal: true

require "bundler/gem_tasks"

begin
  require "rspec/core/rake_task"
  RSpec::Core::RakeTask.new(:spec)
  desc "alias test task to spec"
  task test: :spec
rescue LoadError
  warn("Failed to load rspec")
end

begin
  require "reek/rake/task"
  Reek::Rake::Task.new do |t|
    t.fail_on_error = true
    t.verbose = false
    t.source_files = "lib/**/*.rb"
  end
rescue LoadError
  warn("reek is not installed")
end

require_relative "lib/sanitize_email/version"
require "rdoc"
require "rdoc/task"
RDoc::Task.new do |rdoc|
  rdoc.rdoc_dir = "rdoc"
  rdoc.title = "SanitizeEmail #{SanitizeEmail::Version::VERSION}"
  rdoc.options << "--line-numbers"
  rdoc.rdoc_files.include("README*")
  rdoc.rdoc_files.include("lib/**/*.rb")
end

begin
  require "rubocop/lts"
  Rubocop::Lts.install_tasks
rescue LoadError
  puts "Linting not available"
end

task default: %i[spec rubocop_gradual reek]
```

ref: https://github.com/pboling/sanitize_email/blob/main/Rakefile

**Comment by pboling** [comment]
Lastly, none if this should matter for a released gem, because neither `Rakefile` nor `Gemfile` should ship as part of a gem.  They are because of the glob in the gemspec, but you can replace it with an explicit list.  Here's what I use in my gems now...

Note that the glob only targets things inside `lib/`, and literally everything else is a non-executable text-file:
```
  spec.files = Dir[
    "lib/**/*.rb",
    "CODE_OF_CONDUCT.md",
    "CHANGELOG.md",
    "CONTRIBUTING.md",
    "LICENSE.txt",
    "README.md",
    "SECURITY.md"
  ]
```

Ref: https://github.com/pboling/sanitize_email/blob/main/sanitize_email.gemspec#L27-L35

**Comment by janlelis** [comment]
Thanks for the additional details. I have updated the Rakefile (with [my current default template](https://github.com/janlelis/microgem/blob/main/template/Rakefile.tt) + the RSpec section from the Rakefile you've posted) and dropped rubygems-tasks.

But, I am not convinced that the `Rakefile` shouldn't be shipped, since it helps people that might want to work on the gem. What would be the main reason to drop it (lower package size)?

**Comment by janlelis** [comment]
v2.0.3 is released - happy to hear if it works now

**Comment by pboling** [comment]
The shipped gem is installed in a way that people shouldn't be working on it ever, at all, in that form.  Most IDEs will give you a stern warning if you try to edit an installed "library" file, as it would be just like editing the Ruby source code, and because it isn't in source control it is very hard to undo those edits.  It they want to work on it they should clone the project from GitHub.

If you want to ship a gem that has rake tasks which users of the gem can run, that is done with a `rakelib` directory.

But it sounds like there is some confusion here, you definitely don't want people running specs for the gem via the installed gem.  That is not a use case that is normal or expected anywhere.

Shipping a Rakefile for people who may want to develop the gem is just confusing, as it is entirely the wrong way to develop contributions to a gem.

Having a smaller package size is nice, and in that spirit it _really_ doesn't make sense to glob include all the root files.

**Comment by pboling** [comment]
Still failing, same error as before.
https://github.com/pboling/sanitize_email/actions/runs/8790374551/job/24122248356

**Comment by pboling** [comment]
Perhaps the next step here should be to try to repro in the CI of this gem...
I see this warning in the MacOS 2.5 build, which might be related in that perhaps it is only broken on newer versions of RubyGems / bundler:
```
Your RubyGems version (2.7.6.3) has a bug that prevents `required_ruby_version` from working for Bundler. Any scripts that use `gem install bundler` will break as soon as Bundler drops support for your Ruby version. Please upgrade RubyGems to avoid future breakage and silence this warning by running `gem update --system 3.2.3`
```

**Comment by janlelis** [comment]
> The shipped gem is installed in a way that people shouldn't be working on it ever, at all, in that form. Most IDEs will give you a stern warning if you try to edit an installed "library" file, as it would be just like editing the Ruby source code, and because it isn't in source control it is very hard to undo those edits. It they want to work on it they should clone the project from GitHub.
> 
> If you want to ship a gem that has rake tasks which users of the gem can run, that is done with a `rakelib` directory.
> 
> But it sounds like there is some confusion here, you definitely don't want people running specs for the gem via the installed gem. That is not a use case that is normal or expected anywhere.
> 
> Shipping a Rakefile for people who may want to develop the gem is just confusing, as it is entirely the wrong way to develop contributions to a gem.
> 
> Having a smaller package size is nice, and in that spirit it _really_ doesn't make sense to glob include all the root files.

Thanks for your thoughts on this. My reply (which draws from my own experience moving through the ecosystem) is that I actually do modify installed libraries and packages from time to time for debugging purpose, fixing bugs, or even just to learn more about their implementation... Also, since the RubyGems platform is independent from GitHub, not every project can be found there... (most have an VCS copy hosted somewhere, but it's not a requirement...)

**Comment by janlelis** [comment]
> Perhaps the next step here should be to try to repro in the CI of this gem... I see this warning in the MacOS 2.5 build, which might be related in that perhaps it is only broken on newer versions of RubyGems / bundler:
> 
> ```
> Your RubyGems version (2.7.6.3) has a bug that prevents `required_ruby_version` from working for Bundler. Any scripts that use `gem install bundler` will break as soon as Bundler drops support for your Ruby version. Please upgrade RubyGems to avoid future breakage and silence this warning by running `gem update --system 3.2.3`
> ```

I have added a test build to the CI and it seems that [the gem can be installed successfully](https://github.com/janlelis/ruby_engine/actions/runs/8799221637/job/24151410474). It's not run through bundler, though... Does the hint to run `gem update --system` solve the problem?

**Comment by pboling** [comment]
The issue is upgrading on Ruby 2.5 means upgrading to an older version pinned rubygems, since they dropped support for 2.5 a while ago.
```
gem install rubygems-update -v 3.3.27
```

I do also frequently modify gem's source code when debugging, but if I want to contribute I always find the source, check it out and submit the PR / patch against the VCS.  It's not feasible to even create a patch from modifications to non-version-controlled files, IMO.  Too many pitfalls.

**Comment by janlelis** [comment]
@pboling Could you parse the `DEBUG=debug`env variable to `bundle install`? this way we should get a stacktrace

https://github.com/rubygems/bundler/blob/f66c3346733afeeff3ac4b09f522fe40bc8dbb44/lib/bundler/installer/gem_installer.rb#L32

**Comment by pboling** [comment]
Not quite to the point of debugging this, but in debugging my own CI failures... I found

> If you're running into ArgumentError: wrong number of arguments (given 4, expected 1) errors with a stacktrace including Psych and RubyGems entries, you should be able to fix them by setting rubygems: 3.0.0 or higher.

😆 

I am nearly certain this is the issue, and also certain that I can fix it myself.  Closing.

Ref: https://github.com/ruby/setup-ruby?tab=readme-ov-file#rubygems
